### PR TITLE
SelectionList keyboard navigation

### PR DESCRIPTION
**Pull request by alisey**
I had to replace itemRenderer with renderItem function to be able to read the value attribute from the item node's props. Otherwise we would have to read it from DOM.

### CODE REVIEW

File: src/components/selection-list/selection-list.st.css [hunk 1]
```diff
@@ -1,13 +1,15 @@
...
+    -st-states: focused;
     width: 276px;
-    border: value(borderFocus);
+    border: 1px solid #adbbc6;
+    outline: none;
+}
+
+.divider {
+    height: 1px;
+    background-color: #adbbc6;
```
Comment (idoros) [inline src/components/selection-list/selection-list.st.css]: colors shouldn't come from default theme?

Comment (alisey) [inline src/components/selection-list/selection-list.st.css]: @idoros Right, there are probably no styles that we should apply to the component by default. I started refactoring styling, and then got stuck with trying to fix other related problems without deviating from the spec too much.

File: src/components/selection-list/selection-list.tsx [hunk 2]
```diff
@@ -151,16 +168,34 @@ export class SelectionList extends React.Component<SelectionListProps, {}> {
...
             );
         }
         return child;
     }
 
-    private handleClick: React.EventHandler<React.MouseEvent<HTMLElement>> = event => {
-        const listRoot = ReactDOM.findDOMNode(this);
+    // The `data-value` attribute on the HTML element returned by the itemRenderer is the only source of truth
+    // about the item's value, and we can only get it from DOM :(
+    private getItemValuesFromDOM() {
```
Comment (JoMarton) [inline src/components/selection-list/selection-list.tsx]: what is the returned type?

File: src/components/selection-list/selection-list.tsx [hunk 3]
```diff
@@ -151,16 +168,34 @@ export class SelectionList extends React.Component<SelectionListProps, {}> {
...
+        const rootNode = ReactDOM.findDOMNode(this);
+        const children = Array.from(rootNode.children) as HTMLElement[];
+        return children.map(elem => elem.dataset.value).filter(value => value !== undefined);
+    }
+
+    private handleFocus: React.FocusEventHandler<HTMLElement> = event => {
+        this.setState({focused: true, focusedValue: this.props.value});
```
Comment (JoMarton) [inline src/components/selection-list/selection-list.tsx]: seems to me like this is untested? can we find a way to test it?

File: src/components/selection-list/selection-list.tsx [hunk 4]
```diff
@@ -151,16 +168,34 @@ export class SelectionList extends React.Component<SelectionListProps, {}> {
...
+    }
+
+    private handleBlur: React.FocusEventHandler<HTMLElement> = event => {
+        this.setState({focused: false, focusedValue: undefined});
```
Comment (JoMarton) [inline src/components/selection-list/selection-list.tsx]: same as for `handleFocus`

File: src/components/selection-list/selection-list.tsx [hunk 5]
```diff
@@ -171,5 +206,53 @@ export class SelectionList extends React.Component<SelectionListProps, {}> {
...
+    }
+
+    private handleKeyDown: React.KeyboardEventHandler<HTMLElement> = event => {
+        const values = this.getItemValuesFromDOM();
+        const focusedValue = this.state.focusedValue || this.props.value;
+        const focusedIndex = (focusedValue === undefined) ? -1 : values.indexOf(focusedValue);
+        const maxIndex = values.length - 1;
+
+        switch (event.keyCode) {
+            case keycode('enter'):
```
Comment (giladgrush) [inline src/components/selection-list/selection-list.tsx]: Please remove, 'enter' handling, my bad.

File: src/components/selection-list/selection-list.tsx (hunk 4 above)
Comment (alisey) [inline src/components/selection-list/selection-list.tsx]: @JoMarton added tests for focused/selected states.

File: src/components/selection-list/selection-list.tsx [hunk 6]
```diff
@@ -84,92 +96,166 @@ export interface SelectionListProps extends OptionList {
...
+    focused: boolean;
+    focusedValue?: string;
+}
+
+@SBComponent(listStyle)
+export class SelectionList extends React.Component<SelectionListProps, SelectionListState> {
     public static defaultProps: SelectionListProps = {
         dataSource: [],
-        itemRenderer: DefaultItemRenderer,
+        renderItem,
```
Comment (JoMarton) [inline src/components/selection-list/selection-list.tsx]: give default `tabIndex`

File: src/components/selection-list/selection-list.tsx [hunk 7]
```diff
@@ -84,92 +96,166 @@ export interface SelectionListProps extends OptionList {
...
+
     public render() {
         const rootProps = root(this.props, {
             'data-automation-id': 'LIST',
             'className': 'list',
-            'onClick': this.handleClick
+            'cssStates': {
+                focused: this.state.focused
+            },
+            'tabIndex': -1,
```
Comment (JoMarton) [inline src/components/selection-list/selection-list.tsx]: `tabIndex` should be from props.
`on-*` behavior should be examined, if not overriden by user - put them on `root`

**Pull request merged**